### ISSUE TITLE
Fehler CS0117	'"LightGbmRegressionTrainer.Options" enthält keine Definition für "UseZeroAsMissingValue"

### ISSUE THREAD

**Issue by NebuDa** [dotnet/machinelearning-modelbuilder#719]
**System Information (please complete the following information):**
 - Model Builder Version: 16.0.2003.302
 - Visual Studio Version: Enterprise 2019
                                       Version 16.5.4
 

**Describe the bug**
I Trained the  Model as a Price Prediction Scenario (predict numeric value) with a csv File and added the Projects to my Project. Then I tried to run the ConsoleApp, but gets this Compiler time error:
Fehler	CS0117	'"LightGbmRegressionTrainer.Options" enthält keine Definition für "UseZeroAsMissingValue".	MaschinenWartungsVorhersageML.ConsoleApp	[...path...]\MaschinenWartungsVorhersageML.ConsoleApp\ModelBuilder.cs	53

The automated generated Line of Code:
var trainer = mlContext.Regression.Trainers.LightGbm(new LightGbmRegressionTrainer.Options() { NumberOfIterations = 200, LearningRate = 0.07174045f, NumberOfLeaves = 45, MinimumExampleCountPerLeaf = 10, UseCategoricalSplit = true, HandleMissingValue = false, **UseZeroAsMissingValue = false**, MinimumExampleCountPerGroup = 50, MaximumCategoricalSplitPointCount = 8, CategoricalSmoothing = 10, L2CategoricalRegularization = 0.5, Booster = new GradientBooster.Options() { L2Regularization = 0, 5, L1Regularization = 1 }, LabelColumnName = "Zeit bis Wartung", FeatureColumnName = "Features" });

Output of Machine Learning:
Code Generated
|     Trainer                             RSquared Absolute-loss Squared-loss RMS-loss  Duration #Iteration      |
|1    SdcaRegression                        0,9966          0,99         3,49     1,87       1,9          1      |
|2    LightGbmRegression                    0,9999          0,17         0,07     0,27       1,4          2      |
|3    FastTreeRegression                    0,9999          0,13         0,06     0,25       2,2          3      |
|4    FastTreeTweedieRegression             0,9987          0,32         1,09     1,04       2,0          4      |
|5    FastForestRegression                  0,9981          1,09         1,75     1,32       2,7          5      |
Poisson regression must regress to a non-negative label, but label -0.259 encountered
|6    OnlineGradientDescentRegression       0,9523          5,23        44,39     6,66       0,6          6      |
Input matrix was not positive-definite. Try using a larger L2 regularization weight.
Parametername: ap
|7    LightGbmRegression                    0,9999          0,19         0,10     0,32       1,1          7      |
|8    FastTreeRegression                    0,3757         21,51       643,66    25,37       2,8          8      |
|9    FastTreeTweedieRegression            -0,6698         34,45      1721,52    41,49       1,4          9      |
|10   LightGbmRegression                    0,9999          0,12         0,05     0,22       1,5         10      |
|11   FastTreeRegression                    0,9857          3,27        13,98     3,74       1,0         11      |
|12   FastTreeTweedieRegression             0,9986          0,41         1,13     1,07       3,0         12      |
|13   LightGbmRegression                    0,9998          0,27         0,23     0,48       0,9         13      |
|14   FastTreeRegression                    0,2741         23,21       748,37    27,36       1,2         14      |
|15   FastTreeTweedieRegression             0,9973          0,84         2,42     1,55       4,1         15      |
|16   LightGbmRegression                    0,9991          0,73         0,86     0,93       1,1         16      |

===============================================Experiment Results=================================================
------------------------------------------------------------------------------------------------------------------
|                                                     Summary                                                    |
------------------------------------------------------------------------------------------------------------------
|ML Task: regression                                                                                             |
|Dataset: [...path...]\TSverschleiss8regelmTage.CSV|
|Label : Zeit bis Wartung                                                                                        |
|Total experiment time : 29,3788881 Secs                                                                         |
|Total number of models explored: 18                                                                             |
------------------------------------------------------------------------------------------------------------------

|                                              Top 5 models explored                                             |
------------------------------------------------------------------------------------------------------------------
|     Trainer                             RSquared Absolute-loss Squared-loss RMS-loss  Duration #Iteration      |
|1    LightGbmRegression                    0,9999          0,12         0,05     0,22       1,5          1      |
|2    FastTreeRegression                    0,9999          0,13         0,06     0,25       2,2          2      |
|3    LightGbmRegression                    0,9999          0,17         0,07     0,27       1,4          3      |
|4    LightGbmRegression                    0,9999          0,19         0,10     0,32       1,1          4      |
|5    LightGbmRegression                    0,9998          0,27         0,23     0,48       0,9          5      |
------------------------------------------------------------------------------------------------------------------

Code Generated
Add MaschinenWartungsVorhersageML.Model

**Comment by NebuDa** [comment]
Removing the wrong options leads to this Error in the same line of code:

Fehler	CS0747	Ungültiger Deklarator des Initialisierermembers.	MaschinenWartungsVorhersageML.ConsoleApp

**Comment by JakeRadMSFT** [comment]
@NebuDa Can you share the generated ModelBuilder.cs?

**Comment by NebuDa** [comment]
// This file was auto-generated by ML.NET Model Builder. 

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Microsoft.ML;
using Microsoft.ML.Data;
using MaschinenWartungsVorhersageML.Model;
using Microsoft.ML.Trainers.LightGbm;

namespace MaschinenWartungsVorhersageML.ConsoleApp
{
    public static class ModelBuilder
    {
        private static string TRAIN_DATA_FILEPATH = @"[...path...]\Maschinenverschleiss\TSverschleiss8regelmTage.CSV";
        private static string MODEL_FILEPATH = @"[...path...]\AppData\Local\Temp\MLVSTools\MaschinenWartungsVorhersageML\MaschinenWartungsVorhersageML.Model\MLModel.zip";
        // Create MLContext to be shared across the model creation workflow objects 
        // Set a random seed for repeatable/deterministic results across multiple trainings.
        private static MLContext mlContext = new MLContext(seed: 1);

        public static void CreateModel()
        {
            // Load Data
            IDataView trainingDataView = mlContext.Data.LoadFromTextFile<ModelInput>(
                                            path: TRAIN_DATA_FILEPATH,
                                            hasHeader: true,
                                            separatorChar: ';',
                                            allowQuoting: true,
                                            allowSparse: false);

            // Build training pipeline
            IEstimator<ITransformer> trainingPipeline = BuildTrainingPipeline(mlContext);

            // Evaluate quality of Model
            Evaluate(mlContext, trainingDataView, trainingPipeline);

            // Train Model
            ITransformer mlModel = TrainModel(mlContext, trainingDataView, trainingPipeline);

            // Save model
            SaveModel(mlContext, mlModel, MODEL_FILEPATH, trainingDataView.Schema);
        }

        public static IEstimator<ITransformer> BuildTrainingPipeline(MLContext mlContext)
        {
            // Data process configuration with pipeline data transformations 
            var dataProcessPipeline = mlContext.Transforms.Conversion.ConvertType(new[] { new InputOutputColumnPair("eingeschaltet", "eingeschaltet") })
                                      .Append(mlContext.Transforms.Text.FeaturizeText("Uhrzeit_tf", "Uhrzeit"))
                                      .Append(mlContext.Transforms.Concatenate("Features", new[] { "eingeschaltet", "Uhrzeit_tf", "Last", "Schwingung", "Belastungsfaktor", "Laufzeit" }));
            // Set the training algorithm 
            var trainer = mlContext.Regression.Trainers.LightGbm(new LightGbmRegressionTrainer.Options() { NumberOfIterations = 200, LearningRate = 0.07174045f, NumberOfLeaves = 45, MinimumExampleCountPerLeaf = 10, UseCategoricalSplit = true, HandleMissingValue = false, MinimumExampleCountPerGroup = 50, MaximumCategoricalSplitPointCount = 8, CategoricalSmoothing = 10, L2CategoricalRegularization = 0.5, Booster = new GradientBooster.Options() { L2Regularization = 0, 5, L1Regularization = 1 }, LabelColumnName = "Zeit bis Wartung", FeatureColumnName = "Features" });
            //var trainer = mlContext.Regression.Trainers.LightGbm(new LightGbmRegressionTrainer.Options() { NumberOfIterations = 200, LearningRate = 0.07174045f, NumberOfLeaves = 45, MinimumExampleCountPerLeaf = 10, UseCategoricalSplit = true, HandleMissingValue = false, UseZeroAsMissingValue = false, MinimumExampleCountPerGroup = 50, MaximumCategoricalSplitPointCount = 8, CategoricalSmoothing = 10, L2CategoricalRegularization = 0.5, Booster = new GradientBooster.Options() { L2Regularization = 0, 5, L1Regularization = 1 }, LabelColumnName = "Zeit bis Wartung", FeatureColumnName = "Features" });


            var trainingPipeline = dataProcessPipeline.Append(trainer);

            return trainingPipeline;
        }

        public static ITransformer TrainModel(MLContext mlContext, IDataView trainingDataView, IEstimator<ITransformer> trainingPipeline)
        {
            Console.WriteLine("=============== Training  model ===============");

            ITransformer model = trainingPipeline.Fit(trainingDataView);

            Console.WriteLine("=============== End of training process ===============");
            return model;
        }

        private static void Evaluate(MLContext mlContext, IDataView trainingDataView, IEstimator<ITransformer> trainingPipeline)
        {
            // Cross-Validate with single dataset (since we don't have two datasets, one for training and for evaluate)
            // in order to evaluate and get the model's accuracy metrics
            Console.WriteLine("=============== Cross-validating to get model's accuracy metrics ===============");
            var crossValidationResults = mlContext.Regression.CrossValidate(trainingDataView, trainingPipeline, numberOfFolds: 5, labelColumnName: "Zeit bis Wartung");
            PrintRegressionFoldsAverageMetrics(crossValidationResults);
        }

        private static void SaveModel(MLContext mlContext, ITransformer mlModel, string modelRelativePath, DataViewSchema modelInputSchema)
        {
            // Save/persist the trained model to a .ZIP file
            Console.WriteLine($"=============== Saving the model  ===============");
            mlContext.Model.Save(mlModel, modelInputSchema, GetAbsolutePath(modelRelativePath));
            Console.WriteLine("The model is saved to {0}", GetAbsolutePath(modelRelativePath));
        }

        public static string GetAbsolutePath(string relativePath)
        {
            FileInfo _dataRoot = new FileInfo(typeof(Program).Assembly.Location);
            string assemblyFolderPath = _dataRoot.Directory.FullName;

            string fullPath = Path.Combine(assemblyFolderPath, relativePath);

            return fullPath;
        }

        public static void PrintRegressionMetrics(RegressionMetrics metrics)
        {
            Console.WriteLine($"*************************************************");
            Console.WriteLine($"*       Metrics for Regression model      ");
            Console.WriteLine($"*------------------------------------------------");
            Console.WriteLine($"*       LossFn:        {metrics.LossFunction:0.##}");
            Console.WriteLine($"*       R2 Score:      {metrics.RSquared:0.##}");
            Console.WriteLine($"*       Absolute loss: {metrics.MeanAbsoluteError:#.##}");
            Console.WriteLine($"*       Squared loss:  {metrics.MeanSquaredError:#.##}");
            Console.WriteLine($"*       RMS loss:      {metrics.RootMeanSquaredError:#.##}");
            Console.WriteLine($"*************************************************");
        }

        public static void PrintRegressionFoldsAverageMetrics(IEnumerable<TrainCatalogBase.CrossValidationResult<RegressionMetrics>> crossValidationResults)
        {
            var L1 = crossValidationResults.Select(r => r.Metrics.MeanAbsoluteError);
            var L2 = crossValidationResults.Select(r => r.Metrics.MeanSquaredError);
            var RMS = crossValidationResults.Select(r => r.Metrics.RootMeanSquaredError);
            var lossFunction = crossValidationResults.Select(r => r.Metrics.LossFunction);
            var R2 = crossValidationResults.Select(r => r.Metrics.RSquared);

            Console.WriteLine($"*************************************************************************************************************");
            Console.WriteLine($"*       Metrics for Regression model      ");
            Console.WriteLine($"*------------------------------------------------------------------------------------------------------------");
            Console.WriteLine($"*       Average L1 Loss:       {L1.Average():0.###} ");
            Console.WriteLine($"*       Average L2 Loss:       {L2.Average():0.###}  ");
            Console.WriteLine($"*       Average RMS:           {RMS.Average():0.###}  ");
            Console.WriteLine($"*       Average Loss Function: {lossFunction.Average():0.###}  ");
            Console.WriteLine($"*       Average R-squared:     {R2.Average():0.###}  ");
            Console.WriteLine($"*************************************************************************************************************");
        }
    }
}

**Comment by NebuDa** [comment]
By the way, training as custom regression scenario is working very well, just as Price Prediction does not work. My problem is solved, but the bug in th price prediction scenario is still present.

**Comment by JakeRadMSFT** [comment]
@NebuDa Can you share your dataset or a portion of it?

We've actually removed the custom option in the newer versions of Model Builder ... so I want to make sure we haven't blocked you from being able to retrain.

**Comment by JakeRadMSFT** [comment]
Closing due to lack of response. Please re-open if you're still having this issue!
